feat: add overdetermined morphing

wangzhe04 opened the pull request:
This pull request implements the following changes to morphing.py:

1.  Add the overdetermined morphing functionality which allows users to input more input as many basis points as possible.
2.  Add the additional input format that allows users to specify coupling for production, decay, and both separately
3.  Add an tutorial Jupyter Notebook for using morphing.py in the directory examples/morphing_example.ipynb
4.  Add additional tests for testing the implemented features in tests/test_morphing.py

alexander-held commented:
Thanks for opening this! You can add the `sympy` dependency to `setup.cfg` here: https://github.com/madminer-tool/madminer/blob/dd223f8462120eb968b753499518dee6f229b052/setup.cfg#L28-L36. That should cause it to be picked up for the tests in continuous integration. The tests are probably going to fail without updates, so we'll also have to look into updating those and testing the new functionality.

There are currently a lot of empty lines getting removed, is this on purpose? It might be best to avoid this in this PR and do it separately in another one if it is a matter of just formatting.

Can you reproduce the issue you saw with `particle`? And do you mean `utils/particle.py` or some other file? If there is a conflict that can appear, it would be good to find a way to avoid this as well.

wangzhe04 commented:
Yes, I can reproduce the error, this is the error message i got after running a new runner file in the docker environement. 

Code: 
import morphing as m
if __name__=="__main__":
    a = m.PhysicsMorpher(parameter_max_power=[2,2])

Error:
root@976930b62f62:/madminer/madminer/madminer/utils# /bin/python3 /madminer/madminer/madminer/utils/non_runner.py
Traceback (most recent call last):
  File "/madminer/madminer/madminer/utils/non_runner.py", line 1, in <module>
    import morphing as m
  File "/madminer/madminer/madminer/utils/morphing.py", line 9, in <module>
    from madminer.models import Benchmark
  File "/usr/local/lib/python3.8/dist-packages/madminer/__init__.py", line 4, in <module>
    from .delphes import DelphesReader
  File "/usr/local/lib/python3.8/dist-packages/madminer/delphes/__init__.py", line 1, in <module>
    from .delphes_reader import DelphesReader
  File "/usr/local/lib/python3.8/dist-packages/madminer/delphes/delphes_reader.py", line 10, in <module>
    from madminer.utils.interfaces.delphes_root import parse_delphes_root_file
  File "/usr/local/lib/python3.8/dist-packages/madminer/utils/interfaces/delphes_root.py", line 11, in <module>
    from particle import Particle
  File "/madminer/madminer/madminer/utils/particle.py", line 4, in <module>
    from particle import Particle
ImportError: cannot import name 'Particle' from partially initialized module 'particle' (most likely due to a circular import) (/madminer/madminer/madminer/utils/particle.py)

I think it can be solved by changing the utils/particle.py to a new name such as utils/particle1.py

alexander-held commented:
The issue you see is due to running from within the `madminer` library, you can resolve it by running your code from elsewhere, then
```python
from madminer.utils import morphing as m

if __name__ == "main":
    a = m.PhysicsMorpher(parameter_max_power=[2, 2])
```
works fine. In order to be able to import the code in this fashion, while still being able to change the `morphing` module and see the effects, you can do an [editable install](https://pip.pypa.io/en/stable/topics/local-project-installs/#editable-installs) e.g. via `pip install -e .` from within the `madminer` repository.

matthewfeickert commented on setup.cfg:
```diff
@@ -34,6 +34,7 @@ install_requires =
     torch>=1.0.0
     uproot3>=3.14.1
     vector>=0.8.4
+    sympy>=1.10.1
```
Is this version of `sympy` known to be the oldest version that will still work? Was part of the `sympy` API that is used in this PR added in `v1.10.1`?

wangzhe04 replied:
It's the version that I used in my environment. I assume there are older versions that can work, but haven't tested yet. I'll try to test it sometime later.

matthewfeickert replied:
Yeah it would be good to not be overly restrictive when it comes to core dependencies.

alexander-held commented on madminer/utils/morphing.py:
```diff
@@ -113,80 +119,186 @@ def set_components(self, components):
+        Np : int, optional, default = 0
+            Number of production couplings
+        Ns : int, optional, default = 0
+            Number of spin couplings
```
```suggestion
            Number of same couplings
```
We might need more explanations either here or elsewhere to clarify what this means.

wangzhe04 replied:
Yes I'll add that it stands for both production and decay

Sinclert commented:
👋🏻 Hi @wangzhe04

Thank you for your contributions, they are highly appreciated :)

Tuning in to ask for tests covering the modified [PhysicsMorpher](https://github.com/madminer-tool/madminer/blob/master/madminer/utils/morphing.py) functions before marking this PR ready to review (as of now: `find_components`, `set_basis` and `calculate_morphing_matrix`). The test suite is currently very limited, and increasing it will make development sustainable over time.

wangzhe04 commented:
> 👋🏻 Hi @wangzhe04
> 
> Thank you for your contributions, they are highly appreciated :)
> 
> Tuning in to ask for tests covering the modified [PhysicsMorpher](https://github.com/madminer-tool/madminer/blob/master/madminer/utils/morphing.py) functions before marking this PR ready to review (as of now: `find_components`, `set_basis` and `calculate_morphing_matrix`). The test suite is currently very limited, and increasing it will make development sustainable over time.

Thank you! Thank you for your contributions as well! 
I'll surely include some testings for these modified parts!

matthewfeickert commented:
@alexander-held suggested in a meeting today that @wangzhe04 write some tests for the project that target the areas he's hoping to extend for his IRIS-HEP Fellow project in another PR. 
As @Sinclert has already pointed out, this will give the nice effect of both adding coverage in addition to giving a nice method to check to make sure that the changes @wangzhe04 makes are doing what are expected. :slightly_smiling_face: 

@wangzhe04 if you wanted to outline some of the ideas for this project in a GitHub Issue that might make it easier to step through some other parts later on in the review.

wangzhe04 commented:
> @alexander-held suggested in a meeting today that @wangzhe04 write some tests for the project that target the areas he's hoping to extend for his IRIS-HEP Fellow project in another PR. As @Sinclert has already pointed out, this will give the nice effect of both adding coverage in addition to giving a nice method to check to make sure that the changes @wangzhe04 makes are doing what are expected. 🙂
> 
> @wangzhe04 if you wanted to outline some of the ideas for this project in a GitHub Issue that might make it easier to step through some other parts later on in the review.

Sure, I'll look into creating a issue tomorrow, thanks!

wangzhe04 commented:
Changed to pull request#516

Sinclert commented:
Hey @wangzhe04 .

Thanks again for all the effort you are putting into this. Having tests for such a big change is crucial :)

I am curious though, why are you moving changes to new branches, opening new PRs while closing previous ones? I am afraid conversation / context may be lost because of this _dance_.

wangzhe04 commented:
> Hey @wangzhe04 .
> 
> Thanks again for all the effort you are putting into this. Having tests for such a big change is crucial :)
> 
> I am curious though, why are you moving changes to new branches, opening new PRs while closing previous ones? I am afraid conversation / context may be lost because of this _dance_.

Oh, I saw the changes made in this PR are also showing in the new PR so I thought there is no need to keep this one. I'll reopen this one for reference. Thank you for pointing out.

alexander-held commented:
I think this was a misunderstanding about the strategy, sorry I was probably not very clear about this. My proposal is getting tests for the existing functionality in first (which we can do via #516) and then use this PR to add the new functionality + expand the tests to cover the new additions.

alexander-held commented on madminer/utils/morphing.py:
```diff
@@ -170,23 +252,60 @@ def set_basis(self, basis_from_madminer: Dict[str, Benchmark] = None, basis_nump
+                self.n_benchmarks = len(basis_d[0])
+            # Check the which inputs are provided, and set self.n_benchmarks coordinately. 
+            if basis_s is not None and basis_p is not None and basis_d is not None:
+                assert len(basis_p[0]) == len(basis_d[0]) == len(basis_s[0]), "the number of basis points in production, decay and combine should be the same"
```
`assert` will be disabled when Python is running with some optimizations, so if you want to use this as a way to inform the user, I would suggest raising an exception (e.g. `raise ValueError("some_text")`).

wangzhe04 replied:
Sure, i will change asserts to exceptions

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +129,117 @@ def find_components(self, max_overall_power=4):
             scales a given component.
         """
+        if Nd != 0 or Np != 0 or Ns != 0:
+            lst = []
```
Could you be more precise here? I am not sure what this _list_ is suppose to contain by its name.

wangzhe04 replied:
Thank you for your feedbacks, i'll change it name to components in the next commit.

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -301,7 +420,7 @@ def optimize_basis(
         return best_basis
 
-    def calculate_morphing_matrix(self, basis=None):
+    def calculate_morphing_matrix(self, basis=None, gd = None, gp = None, gs = None):
```
Even if we have not officially added Black as the project format checker, we should try to follow its style guide.

```suggestion
    def calculate_morphing_matrix(self, basis=None, gd=None, gp=None, gs=None):
```

wangzhe04 replied:
Thank you, I didn't run black in this commit but will run it before next commit.

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +129,117 @@ def find_components(self, max_overall_power=4):
+    def set_basis(self, basis_from_madminer: Dict[str, Benchmark] = None, basis_numpy=None, morphing_matrix=None, basis_p = None, basis_d = None, basis_s = None):
         """
         Manually sets the basis benchmarks.
-
```
Please leave this empty line.

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -311,62 +430,142 @@ def calculate_morphing_matrix(self, basis=None):
+            morphing_submatrix = inv_morphing_submatrix.T
+            self.matrix_before_invertion = morphing_submatrix
+            # QR factorization
+            q, r= np.linalg.qr(morphing_submatrix, 'reduced')
```
```suggestion
            q, r = np.linalg.qr(morphing_submatrix, 'reduced')
```

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -396,33 +604,61 @@ def calculate_morphing_weights(self, theta, basis=None, morphing_matrix=None):
+                factor = 1.0
+                for p in range(self.n_parameters):
+                    factor *= float(theta[p] ** self.components[c, p])
+                component_weights[c] = factor
```
This is occurrence number 1 of the _component weights calculation_ snippet.

Please extract these lines to a private method within this class.

wangzhe04 replied:
Sure, I'll extract these to a private method

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -396,33 +604,61 @@ def calculate_morphing_weights(self, theta, basis=None, morphing_matrix=None):
+                factor = 1.0
+                for p in range(self.n_parameters):
+                    factor *= float(theta[p] ** self.components[c, p])
+                component_weights[c] = factor
```
This is occurrence number 2 of the _component weights calculation_ snippet.

Please extract these lines to a private method within this class.

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -5,6 +5,8 @@
 from typing import Dict
 from typing import Iterable
+from sympy import Poly, symbols, expand
+from numpy import Infinity, linalg as la
```
I would rather have Sympy being imported as `sp`, instead of importing specific functions / modules directly. Reason being: `symbols` and `expand` as generic enough words that may collide with variable / function names  in the future.

Not sure if there is any reason why this should not be the case, I am not familiar with this library.

wangzhe04 replied:
yes i think it can be changed to import as sp

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +129,117 @@ def find_components(self, max_overall_power=4):
+                for k in range(1, Nd):
+                    if(arr[j, k] > power_max):
+                        lst_pos.append(j)
+                    break    
```
If this loop is iterating until it finds _\<something\>_, shouldn't it have the `break` clause inside the if?

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +129,117 @@ def find_components(self, max_overall_power=4):
+                for k in range(Nd+1, Nd+Np):
+                    if(arr[j, k] > power_max):
+                        lst_pos.append(j)
+                    break
```
If this loop is iterating until it finds _\<something\>_, shouldn't it have the `break` clause inside the if?

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +129,117 @@ def find_components(self, max_overall_power=4):
+                for k in range(Nd+Np+1, Nd+Np+Ns):
+                    if(arr[j, k] > power_max):
+                        lst_pos.append(j)
+                    break
```
If this loop is iterating until it finds _\<something\>_, shouldn't it have the `break` clause inside the if?

wangzhe04 replied:
Yes, sorry for this mistake. I actually realized it when writing the test, I'll fix it in the next commit

Sinclert reviewed (commented):
I noticed the PR adding Morphing tests (https://github.com/madminer-tool/madminer/pull/516) was directly merged from its _draft_ state, so I decided to go ahead and provide feedback for this one despite its current state.

Once again, thanks for your contributions :)

alexander-held commented:
> I noticed the PR adding Morphing tests (#516) was directly merged from its _draft_ state, so I decided to go ahead and provide feedback for this one despite its current state.

I unmarked from draft before looking over it again and decided to go ahead since it was a straightforward PR in my view. If you prefer, I am happy to always wait for your explicit approval instead @Sinclert. Either way, this PR is bigger and we were planning on sending a request once all changes are done (which should be within the next day or two I think).

wangzhe04 commented:
This new commit should addresses issues mentioned above, please let me know if i missed anything or there are some other issues. 
Thank you all for your feedbacks.

alexander-held commented on examples/morphing_basis/morphing_example.ipynb:
```diff
@@ -0,0 +1,388 @@
+   ],
+   "source": [
+    "\"\"\"\n",
+    "+----+-----------+---------------------+---------------------+\n",
+    "| g2 | simulated |      n_base = 5     |      n_base = 7     |\n",
+    "+----+-----------+---------------------+---------------------+\n",
+    "| -5 |   0.759   |  0.7590000033377339 |  0.7590043565724045 |\n",
+    "| -4 |    0.53   |  0.5299999713898771 |  0.5300930421799421 |\n",
+    "| -3 |    0.4    | 0.40000000596042407 | 0.39946969877928495 |\n",
+    "| -2 |   0.335   | 0.33500000834464144 |  0.3361039091832936 |\n",
+    "| -1 |   0.316   |  0.3160000145435327 | 0.31487447768449783 |\n",
+    "| 0  |   0.316   | 0.33900004625320435 |  0.3165692798793316 |\n",
+    "| 1  |   0.328   | 0.41500043869018555 | 0.32788529992103577 |\n",
+    "| 2  |    0.34   |  0.5700044631958008 | 0.34142857789993286 |\n",
+    "| 3  |   0.354   |  0.8450050354003906 | 0.35571449995040894 |\n",
+    "| 4  |   0.364   |  1.2960090637207031 |  0.3751673698425293 |\n",
+    "| 5  |   0.376   |  1.9940032958984375 |  0.4101197123527527 |\n",
+    "| 7  |   0.4205  |   4.49005126953125  |  0.5974075496196747 |\n",
+    "| 9  |   0.5347  |  9.201080322265625  |  1.1184179782867432 |\n",
+    "| 11 |   0.7822  |  17.23541259765625  |  2.2685461044311523 |\n",
+    "| 13 |   1.244   |  29.94012451171875  |  4.4378204345703125 |\n",
+    "+----+-----------+---------------------+---------------------+\n",
+    "\"\"\"\n",
+    "print()"
```
I would suggest converting this to a comment field with a Markdown table since you do not use any of the content directly.

wangzhe04 replied:
Sure, I changed it to the markdown table in the new commit

alexander-held commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +130,130 @@ def find_components(self, max_overall_power=4):
+            len_non_pmax = len(non_pmax_components)
+
+            # cut over power_max
+            power_max = BSM_max_power
```
Is there a reason to reassign this variable name, or can we just keep using `BSM_max_power` here?

wangzhe04 replied:
I changed it to keep using BSM_max_power in the new commit

alexander-held commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +130,130 @@ def find_components(self, max_overall_power=4):
+            The maximal sum of powers of all parameters contributing to the squared matrix element for the BSM couplings.
+
+        Nd : int, optional
+            The number of decay vertices.
```
```suggestion
            The number of parameters affecting the decay vertex.
```

alexander-held commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +130,130 @@ def find_components(self, max_overall_power=4):
+            The number of decay vertices.
+
+        Np : int, optional
+            The number of production vertices.
```
```suggestion
            The number of parameters affecting the production vertex.
```

alexander-held commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +130,130 @@ def find_components(self, max_overall_power=4):
+            The number of production vertices.
+
+        Ns : int, optional
+            The number of couplings that works both for production and decay.
```
```suggestion
            The number of parameters that affect both the production and decay vertices.
```

alexander-held commented on madminer/utils/morphing.py:
```diff
@@ -113,7 +120,7 @@ def set_components(self, components):
         self.n_components = len(self.components)
 
-    def find_components(self, max_overall_power=4):
+    def find_components(self, max_overall_power=4, BSM_max_power=float("inf"), Nd=0, Np=0, Ns=0):
```
Suggestion: I would find an order where `Nd` appears before `Np` more intuitive, as production happens before decay. It does not matter for the code, and users can specify the keyword arguments in the order they like, so I won't insist on this change.

wangzhe04 replied:
Sure, I think I have changed all orders into Np, Nd, Ns in parameter inputs and comments in the new commit.

alexander-held commented on madminer/utils/morphing.py:
```diff
@@ -311,62 +464,165 @@ def calculate_morphing_matrix(self, basis=None):
+             Will be used only when gd, gp, gs are None.
+
+        gd : ndarray or None, optional
+                Manually specified the morphing basis of decay basis for which morphing matrix is calculated.
```
```suggestion
                Manually specified morphing basis of decay basis for which morphing matrix is calculated.
```

alexander-held replied:
(same for the other arguments below)

alexander-held commented on madminer/utils/morphing.py:
```diff
@@ -311,62 +464,165 @@ def calculate_morphing_matrix(self, basis=None):
+        gs=None,
+        gd=None,
+        gp=None,
+        theta_d=None,
```
same comment as above about the order of production/decay

alexander-held commented on examples/morphing_basis/morphing_example.ipynb:
```diff
@@ -0,0 +1,373 @@
+    "# Components is the power of each parameter in the given coupling combination, for example, if the equation is gs_0^2 * gs_1^3, the components for this combination is [2,3]\n",
+    "# if there's no known components, we can use the function find_components() to find the correspondig components based of the number of parameter of basis, which equals len(parameter_max_power)\n",
+    "components = morpher_1.find_components()\n",
+    "# print(\"Generated Components: \\n\", components)\n",
```
Please either uncomment this, or remove the lines if you think the information is not particularly interesting to put in the notebook.

wangzhe04 replied:
I uncomment the printout in the new commit. I originally thought leaving this can let user to choose to check and compare the results

alexander-held commented on examples/morphing_basis/morphing_example.ipynb:
```diff
@@ -0,0 +1,373 @@
+    "# First, we should create a new morpher object using the code below\n",
+    "morpher_1 = m.PhysicsMorpher(parameter_max_power=[2, 2])\n",
+    "\n",
+    "# Components is the power of each parameter in the given coupling combination, for example, if the equation is gs_0^2 * gs_1^3, the components for this combination is [2,3]\n",
```
Please try to keep the line lengths of these comments a bit shorter to avoid rendering issues.

wangzhe04 replied:
thank you, I had reduced the lengths of many comments.

alexander-held commented on examples/morphing_basis/morphing_example.ipynb:
```diff
@@ -0,0 +1,373 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "First Input Format"
```
You can turn this into a title by adding a `#` in front, I would suggest giving the whole notebook a title like `# Morphing examples` and then making this a subtitle like `## First Input Format`.

wangzhe04 replied:
Thank you, I added the general title and changed some texts to subtiltes

alexander-held commented:
@wangzhe04 please also edit the very first message in this PR to include a brief high level summary of the changes here.

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +130,127 @@ def find_components(self, max_overall_power=4):
+            # list of tuples containing monomials
+
+            for i in range(0, len(mono)):
+                components.append(mono[i][0])
```
Three comments here:

1. The iterator can be simplified: `range(0, len(...))` --> `range(len(...))` --> `enumerate(...)`
2. If you are iterating on an array just to access its elements (i.e. `mono[i]`), iterate through them directly
3. It seems the `components` array is not used, anywhere, before this point. Why not creating it here?

```suggestion
            components = [mono_value[0] for mono_value in mono]
```

wangzhe04 replied:
Thank you for the feedbacks! I've made the changes and also move the components down here.

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -123,41 +130,127 @@ def find_components(self, max_overall_power=4):
+            exceed_pos = []
+
+            # Find the positions of the subarray that has elements exceed power_max
+            for j in range(0, len_non_pmax):
```
(Do not forget to remove variable `len_non_pmax` if this suggestion gets applied).

Similar to the feedback provided in a previous comment:

```suggestion
            for j, _ in enumerate(non_pmax_components):
```

wangzhe04 replied:
Thank you, I have remove the variable len_non_pmax.

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -573,16 +988,30 @@ def _propose_basis(self, fixed_benchmarks, n_missing_benchmarks):
+            component_weights[c] = factor
+        component_weights = np.array(component_weights)
+
+        return component_weights
```
No need to separate `np.array` conversion, and `return` clause in 2 different lines.

```suggestion
        return np.array(component_weights)
```

Sinclert replied:
@wangzhe04 remember to delete [this line](https://github.com/wangzhe04/madminer/blob/d3caf7971155cda1cb7a23330e86ecce4ac4fd99/madminer/utils/morphing.py#L1009) above the one I suggested, as it is now useless.

wangzhe04 replied:
Thanks for reminding me!, I have deleted in the new commit.

wangzhe04 commented:
In the new commit I included a function in the example notebook that it can check the minimum number of inputs requires. Should we include that in the morphing.py as well? It's a simple function so probably can be added the easily

alexander-held commented:
That looks very useful to me! I think this would be a great addition to `morphing.py` in a new function (with a dedicated test for just this piece of functionality).

wangzhe04 commented:
I added get_min_basis function and the tests and also updated the morphing_example.ipynb to use the built in one. There was a mistake and some format issues in the test_morphing.py i first commited but it should be fixed now.

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -385,44 +645,114 @@ def calculate_morphing_weights(self, theta, basis=None, morphing_matrix=None):
+                            if n_d != 0:
+                                factor *= float(theta_p[i] ** self.components[c, i + n_d])
+                            else:
+                                factor *= float(theta_p[i] ** self.components[c, i])
```
This `if` - `else` block can be simplified. If you are adding the value `n_d` to `i` only when it is not zero, you might as well do it always, even if it is zero (_innocuous operation_) 🤷🏻‍♂️

```python
factor *= float(theta_p[i] ** self.components[c, i + n_d])
```

wangzhe04 replied:
Thank you for your feedback! I made this change and the below one in the new commit

Sinclert commented on madminer/utils/morphing.py:
```diff
@@ -385,44 +645,114 @@ def calculate_morphing_weights(self, theta, basis=None, morphing_matrix=None):
+                            elif n_p != 0:
+                                factor *= float(theta_s[k] ** self.components[c, k + n_p])
+                            else:
+                                factor *= float(theta_s[k] ** self.components[c, k])
```
This whole `if` - `elif` -  `else` block can be simplified. If you are adding the values `n_d` and `n_p` to `k` only when they are not zero, you might as well do it always, even if they are zero (_innocuous operation_) 🤷🏻‍♂️

```python
factor *= float(theta_s[k] ** self.components[c, k + n_d + n_p])
```

Sinclert commented on tests/test_morphing.py:
```diff
@@ -105,3 +105,274 @@ def test_original_weight_gradients(
+                -1.0,
+                1.0,
+            ),
+            (-1.0, 1.0),
```
Could you keep tuple style consistent?

```python
parameter_range=[
     (-1.0, 1.0),
     (-1.0, 1.0),
]
```

Sinclert commented on tests/test_morphing.py:
```diff
@@ -105,3 +105,274 @@ def test_original_weight_gradients(
+                -1.0,
+                1.0,
+            ),
+            (-1.0, 1.0),
```
Keep tuple style consistent:

```python
parameter_range=[
     (-1.0, 1.0),
     (-1.0, 1.0),
]
```

wangzhe04 replied:
Thank you for your feedback, I changed these two input format in the new commit.

Sinclert reviewed (commented):
This is probably the last batch of requested changes! I will approve this PR as soon as they are addressed.

Sinclert reviewed (approved):
LGTM as it is! What do you think? @matthewfeickert @alexander-held

alexander-held reviewed (approved):
This looks good to me, I have no other comments. Thank you very much for getting this ready and all the interactions on the PR!

matthewfeickert reviewed (approved):
Taking a very quick look at this before lunch I think this is fine (added tests are very nice!). If I have any other comments I will open them up as separate Issues that can get resolved after this is merged. Thanks for the contributions @wangzhe04 and for @Sinclert and @alexander-held for the guidance here. :+1:

alexander-held commented:
Thanks everyone! I'll leave it to @Sinclert to hit merge. For future reference, @wangzhe04 will present this work done as an IRIS-HEP fellow also at a meeting tomorrow: https://indico.cern.ch/event/1195271/.
